ground/src/main/java/com/google/android/ground/ui/map/gms/mog/MogClient.kt

```diff
@@ -130,52 +126,57 @@ class MogClient(val collection: MogCollection) {
   }
 
   /** Returns metadata for the MOG bounded by the specified web mercator tile. */
-  private suspend fun getMogMetadata(url: String, bounds: TileCoordinates): MogMetadata? =
-    getMogMetadataAsync(url, bounds).await()
+  private suspend fun getMogMetadata(path: String, bounds: TileCoordinates): MogMetadata? =
+    getMogMetadataAsync(path, bounds).await()
 
   /**
-   * Returns a future containing the [MogMetadata] with the specified extent and URL. Metadata is
+   * Returns a future containing the [MogMetadata] with the specified extent and path. Metadata is
    * either loaded from in-memory cache if present, or asynchronously fetched if necessary. The
    * process of checking the cache and creating the new job is synchronized on the current instance
    * to prevent duplicate parallel requests for the same resource.
    */
-  private fun getMogMetadataAsync(url: String, mogBounds: TileCoordinates): Deferred<MogMetadata?> =
-    synchronized(this) { cache.get(url) ?: getMogMetadataFromRemoteAsync(url, mogBounds) }
+  private fun getMogMetadataAsync(
+    path: String, mogBounds: TileCoordinates
+  ): Deferred<MogMetadata?> =
+    synchronized(this) { cache.get(path) ?: getMogMetadataFromRemoteAsync(path, mogBounds) }
 
   /**
-   * Asynchronously fetches and returns the [MogMetadata] with the specified extent and URL. The
+   * Asynchronously fetches and returns the [MogMetadata] with the specified extent and path. The
    * async job is added to the cache immediately to prevent duplicate fetches from other threads.
    */
   private fun getMogMetadataFromRemoteAsync(
-    url: String,
-    mogBounds: TileCoordinates
+    path: String, mogBounds: TileCoordinates
   ): Deferred<MogMetadata?> = runBlocking {
     // TODO: Exceptions get propagated as cancellation of the coroutine. Handle them!
     async {
-        nullIfNotFound { UrlInputStream(url) }?.use { readMogMetadataAndClose(url, mogBounds, it) }
+      Timber.v("Download path: %s", path)
+      var url: String? = path
+      if (path.startsWith("/")) {
+        // Assume relative path should be resolved by the remote storage manager.
+        url = nullIfNotFound { remoteStorageManager.getDownloadUrl(path).toString() }
       }
-      .also { cache.put(url, it) }
+      Timber.v("Download URL: %s", url)
+      if (url == null) {
+       null
+      } else {
+       nullIfNotFound { UrlInputStream(url) }?.use { readMogMetadataAndClose(url, mogBounds, it) }
```

Comment:
To make that helpful fun more kotlinic, you could do something like this:

```kt
typealias MogPath = String
typealias MogUrl = String
```

Then you could define the helpers as extension member functions so they're scoped to `this`:

```kt
fun MogPath.toUrl(): MogUrl =
  if (startsWith("/")) {
    nullIfNotFound { 
  remoteStorageManager.getDownloadUrl(path).toString() }
        }
  } else this

fun MogUrl.readMetadata(mogBounds: TileCoordinates): MogMetadata? =
  nullIfNotFound { UrlInputStream(this@MogUrl) }?.use { 
    readMogMetadataAndClose(this@MogUrl, mogBounds, it)
  }
```

`@MogUrl` may not be needed here.

Then, this function becomes super simple:

```kt
async {
  getGetMogUrlFromPath(path)?.readMetadata(mogBounds)
}
```

Wdyt?

Reply:
@sufyanAbbasi Made a few corrections in this code example, PTAL!

Reply:
Note you could even give `readMogMetadataAndClose` a similar treatment by defining it as a member fun `fun MogUrl.readMetadataAndClose(mogBounds, inputStream)`

Reply:
Thank you so much for the suggestions, this is brilliant! Done.